sample/samplers.go

```diff
@@ -2,43 +2,88 @@ package sample
 
 import (
 	"errors"
+	"math"
 	"math/rand/v2"
 	"slices"
-)
+	"sync"
 
-// Sampler is not thread-safe. Each goroutine should have its own instance
-type Sampler interface {
-	Sample([]float32) (int32, error)
-}
+	"github.com/ollama/ollama/llama"
+)
 
-// logit represents information about a single token during sampling
-type logit struct {
+// token represents information about a single token during sampling
+type token struct {
 	id    int32   // The token's unique identifier
 	value float32 // The raw logit or probability from the model
 }
 
-type weighted struct {
+type Sampler struct {
 	rng         *rand.Rand
-	tokens      []logit
 	topK        int
 	topP        float32
 	minP        float32
 	temperature float32
+	grammar     *Grammar
 }
 
-func (s *weighted) Sample(logits []float32) (int32, error) {
-	if len(s.tokens) < len(logits) {
-		s.tokens = make([]logit, len(logits))
+func (s *Sampler) Sample(logits []float32) (int32, error) {
+	tokens := make([]token, len(logits))
+	for i := range logits {
+		tokens[i].id = int32(i)
+		tokens[i].value = logits[i]
 	}
 
-	tokens := s.tokens[:len(logits)]
+	t, err := s.sample(tokens)
+	if err != nil {
+		return -1, err
+	}
 
-	for i, v := range logits {
-		tokens[i].id = int32(i)
-		tokens[i].value = v
+	if s.grammar != nil {
+		// optimization: first check if the max logit is accepted by the grammar
+		// if the max logit is rejected, apply the grammar to all logits (slower)
+		top := []token{t}
+		s.grammar.Apply(top)
+		if !math.IsInf(float64(top[0].value), -1) {
+			s.grammar.Accept(top[0].id)
+			return top[0].id, nil
+		}
+
+		// since .sample has side effects of modifying the tokens
+		// we need to reset them before applying the grammar and
+		// sampling again
+		for i := range logits {
+			tokens[i].id = int32(i)
+			tokens[i].value = logits[i]
+		}
+		s.grammar.Apply(tokens)
+		t, err = s.sample(tokens)
+		if err != nil {
+			return -1, err
+		}
+		s.grammar.Accept(t.id)
+	}
+
+	return t.id, nil
+}
+
+func (s *Sampler) greedy(tokens []token) (t token) {
+	max := tokens[0].value
+	for i := 1; i < len(tokens); i++ {
+		if tokens[i].value > max {
+			t = tokens[i]
+			max = t.value
+		}
+	}
+
+	return t
+}
```

Comment:
bug where index 0 can never be returned - fix below. Also cool you can have named return vars that's cool 

```suggestion
	t := tokens[0].value
	for i := 1; i < len(tokens); i++ {
		if tokens[i].value > max {
			t = tokens[i]
		}
	}

	return t
}
```